### PR TITLE
System user, local RPC, demos use RPC

### PR DESCRIPTION
**Pull request by exFalso**
This is a reopen from bitbucket. To sum up:

This PR is the first step to factoring out of the webserver into a separate process. This requires plugin web apis to not depend on the internal ServiceHub, but instead use the RPC interface. As web auth is not yet figured out a god-mode user was added for the webserver.

It also required some code moving, most importantly the RPC interface was moved from `:node` to `:core`

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt [hunk 1]
```diff
@@ -84,8 +91,41 @@ class CordaRPCOpsImpl(
...
+    override fun nodeTimeInUTC(): LocalDateTime = LocalDateTime.now(services.clock)
+
+    override fun partyFromKey(key: CompositeKey) = services.identityService.partyFromKey(key)
+    override fun partyFromName(name: String) = services.identityService.partyFromName(name)
+
+    companion object {
+        fun stateMachineInfoFromFlowStateMachineImpl(stateMachine: FlowStateMachineImpl<*>): StateMachineInfo {
+            return StateMachineInfo(
+                    id = stateMachine.id,
+                    flowLogicClassName = stateMachine.logic.javaClass.simpleName,
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: Perhaps document this property that it's not the fully-qualified class name but rather the class simple name.

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt [hunk 2]
```diff
@@ -160,6 +143,6 @@ inline fun <T : Any, A, B, C, D, reified R : FlowLogic<T>> CordaRPCOps.startFlow
 
 data class FlowHandle<A>(
         val id: StateMachineRunId,
-        val progress: Observable<ProgressTracker.Change>,
+        val progress: Observable<String>,
         val returnValue: Observable<A>
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: As a side point, I disagree that this should be an Observable when we know it will emit only one value. ListenableFuture better captures this intent. WDYT?

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt [hunk 3]
```diff
@@ -160,6 +143,6 @@ inline fun <T : Any, A, B, C, D, reified R : FlowLogic<T>> CordaRPCOps.startFlow
 
 data class FlowHandle<A>(
         val id: StateMachineRunId,
-        val progress: Observable<ProgressTracker.Change>,
+        val progress: Observable<String>,
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: This was a nice type-safe `Observable` of `Change`s. Why the change?

File: node/src/main/kotlin/net/corda/node/services/RPCUserService.kt [hunk 4]
```diff
@@ -11,22 +13,44 @@ import net.corda.node.services.config.FullNodeConfiguration
 interface RPCUserService {
     fun getUser(username: String): User?
     val users: List<User>
+
+    /**
+     * TODO The system user is used by the web api until we implement web auth properly
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/RPCUserService.kt]: Can you update this doc to make it explicit that this is a temporary hack.

File: node/src/main/kotlin/net/corda/node/services/RPCUserService.kt [hunk 5]
```diff
@@ -11,22 +13,44 @@ import net.corda.node.services.config.FullNodeConfiguration
...
+     * The password should be unique per process, so in theory only the node itself should have access to it.
+     */
+    fun addSystemUserPermission(permission: String)
+    companion object {
+        val systemUserUsername = "~systemUser"
+        val systemUserPassword: String = SecureHash.randomSHA256().toString()
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/RPCUserService.kt]: Explicit type isn't needed

File: node/src/main/kotlin/net/corda/node/services/RPCUserService.kt [hunk 6]
```diff
@@ -11,22 +13,44 @@ import net.corda.node.services.config.FullNodeConfiguration
...
 
 // TODO Store passwords as salted hashes
 // TODO Or ditch this and consider something like Apache Shiro
 class RPCUserServiceImpl(config: FullNodeConfiguration) : RPCUserService {
 
-    private val _users = config.rpcUsers.associateBy(User::username)
+    private val _users: HashMap<String, User>
+
+    init {
+        val defaultUsers = listOf(User(RPCUserService.systemUserUsername, RPCUserService.systemUserPassword, setOf()))
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/RPCUserService.kt]: Can you statically import the these two - the line looks a little cluttered

File: node/src/main/kotlin/net/corda/node/services/messaging/RPCStructures.kt [hunk 7]
```diff
@@ -92,7 +78,7 @@ val CURRENT_RPC_USER: ThreadLocal<User> = ThreadLocal()
 /** Helper method which checks that the current RPC user is entitled for the given permission. Throws a [PermissionException] otherwise. */
 fun requirePermission(permission: String) {
     if (permission !in CURRENT_RPC_USER.get().permissions) {
-        throw PermissionException("User not permissioned for $permission")
+        throw PermissionException("User not permissioned for $permission, permissions are ${CURRENT_RPC_USER.get().permissions}")
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/messaging/RPCStructures.kt]: Extract out the current `User` into a local variable

File: samples/irs-demo/src/main/kotlin/net/corda/irs/plugin/IRSPlugin.kt [hunk 8]
```diff
@@ -2,25 +2,26 @@ package net.corda.irs.plugin
...
 import java.time.Duration
 
 class IRSPlugin : CordaPluginRegistry() {
-    override val webApis: List<Class<*>> = listOf(InterestRateSwapAPI::class.java)
+    override val webApis = listOf(::InterestRateSwapAPI)
     override val staticServeDirs: Map<String, String> = mapOf(
             "irsdemo" to javaClass.classLoader.getResource("irsweb").toExternalForm()
     )
-    override val servicePlugins: List<Class<*>> = listOf(FixingFlow.Service::class.java)
+    override val servicePlugins = listOf({ services: PluginServiceHub -> FixingFlow.Service(services) })
```
Comment (shamsasari) [inline samples/irs-demo/src/main/kotlin/net/corda/irs/plugin/IRSPlugin.kt]: `FixingFlow::Service` should work here right?

File: samples/raft-notary-demo/src/main/kotlin/net/corda/notarydemo/flows/DummyIssueAndMove.kt [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package net.corda.notarydemo.flows
+
+import co.paralleluniverse.fibers.Suspendable
+import net.corda.core.contracts.DummyContract
+import net.corda.core.crypto.Party
+import net.corda.core.flows.FlowLogic
+import net.corda.core.node.recordTransactions
+import net.corda.core.transactions.SignedTransaction
+import java.util.*
+
```
Comment (shamsasari) [inline samples/raft-notary-demo/src/main/kotlin/net/corda/notarydemo/flows/DummyIssueAndMove.kt]: Remove extra blank line

File: samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/services/SimmService.kt [hunk 10]
```diff
@@ -17,14 +18,15 @@ import java.time.LocalDate
...
         override val requiredFlows: Map<String, Set<String>> = mapOf(
                 SimmFlow.Requester::class.java.name to setOf(Party::class.java.name, LocalDate::class.java.name),
                 SimmRevaluation.Initiator::class.java.name to setOf(StateRef::class.java.name, LocalDate::class.java.name),
                 IRSTradeFlow.Requester::class.java.name to setOf(SwapData::class.java.name, Party::class.java.name))
-        override val servicePlugins: List<Class<*>> = listOf(
-                SimmFlow.Service::class.java,
-                IRSTradeFlow.Service::class.java)
         override val staticServeDirs: Map<String, String> = mapOf("simmvaluationdemo" to javaClass.classLoader.getResource("simmvaluationweb").toExternalForm())
+        override val servicePlugins: List<(PluginServiceHub) -> Any> = listOf(
+                { services -> SimmFlow.Service(services) },
```
Comment (shamsasari) [inline samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/services/SimmService.kt]: `SimmFlow::Service` and similarly below

File: samples/trader-demo/src/main/kotlin/net/corda/traderdemo/plugin/TraderDemoPlugin.kt [hunk 11]
```diff
@@ -2,17 +2,18 @@ package net.corda.traderdemo.plugin
...
     // A list of classes that expose web APIs.
-    override val webApis: List<Class<*>> = listOf(TraderDemoApi::class.java)
-    // A list of flows that are required for this cordapp
+    override val webApis = listOf(::TraderDemoApi)
+    // A list of Flows that are required for this cordapp
     override val requiredFlows: Map<String, Set<String>> = mapOf(
             SellerFlow::class.java.name to setOf(Party::class.java.name, Amount::class.java.name)
     )
-    override val servicePlugins: List<Class<*>> = listOf(BuyerFlow.Service::class.java)
+    override val servicePlugins = listOf({ services: PluginServiceHub -> BuyerFlow.Service(services) })
```
Comment (shamsasari) [inline samples/trader-demo/src/main/kotlin/net/corda/traderdemo/plugin/TraderDemoPlugin.kt]: And here

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt (hunk 2 above)
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: I think the reason is that the RPC system doesn't know how to transport a ListenableFuture, whereas it does know how to transport observables.

**Comment by mikehearn** [comment]
You might want to rebase on Shams' work, which already introduces various users for node components.

File: core/sandbox/src/main/java/META-INF/MANIFEST.MF [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Manifest-Version: 1.0
+Main-Class: com.r3cev.NonDeter
```
Comment (rick-r3) [inline core/sandbox/src/main/java/META-INF/MANIFEST.MF]: Is this file intentionally here?

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt [hunk 13]
```diff
@@ -112,6 +84,17 @@ interface CordaRPCOps : RPCOps {
      * Retrieve existing note(s) for a given Vault transaction
      */
     fun getVaultTransactionNotes(txnId: SecureHash): Iterable<String>
+
+    fun attachmentExists(id: SecureHash): Boolean
+    fun uploadAttachment(jar: InputStream): SecureHash
+    fun nodeTimeInUTC(): LocalDateTime
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: Good idea.  And convert in the code using the RPC client if we need to.

**Comment by rick-r3** [comment]
It looks like some rogue files have got swept up here, including some zip'd log files.  Please double check and remove them.

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt (hunk 2 above)
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: Wouldn't hurt to put a comment setting expectations of a single observation here.  Generally some docs for ``FlowHandle`` would be good.

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt (hunk 1 above)
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: Will we regret this not being fully qualified later?  The client can always display the truncated name if we wish but hard to add back the package name.

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt (hunk 3 above)
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: It never worked as the serialisers are not registered, we just didn't notice. And trying to register the serialisers is hacky as the ProgressTracker code has some hidden classes.

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt (hunk 2 above)
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: Exactly. We may be able to extend it with some additional functionality, or perhaps even expose a lower level api, but for now only Observables are supported, adding support for something else is out of scope for this pr

File: node/src/main/kotlin/net/corda/node/services/RPCUserService.kt (hunk 5 above)
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/RPCUserService.kt]: IntelliJ gives the usual warning about platform types if it's not specified

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt (hunk 3 above)
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: Correction: the issue wasn't hidden classes, but other nested Observables and backlinks. The `String`s are also consistent with the other way of retrieving progress tracker events (through the state machine rpc). We'll need to rethink the progress tracker code anyway, perpahs then we can think about a more serialisation-friendly data model

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt (hunk 1 above)
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: I think the only place this is used currently is the explorer. I'll change it to be fully qualified. @patrickkuo the ui code may need some adjustment

**Comment by exFalso** [comment]
Addressed all comments, ball in your courts

File: node/src/main/kotlin/net/corda/node/services/RPCUserService.kt (hunk 5 above)
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/RPCUserService.kt]: Ah, `SecureHash.toString` doesn't explicitly specify the return type (because it's calling Java code). Fixing that will remove the need here.

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt (hunk 2 above)
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: Yes - we could extend RPC to support marshalling of futures, but it's a separate project. For now a comment explaining this would be sufficient.

**Comment by mikehearn** [comment]
I have a half-finished review, hold on.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 14]
```diff
@@ -121,6 +124,17 @@ class Node(override val configuration: FullNodeConfiguration, networkMapAddress:
         val legalIdentity = obtainLegalIdentity()
         val myIdentityOrNullIfNetworkMapService = if (networkMapService != null) legalIdentity.owningKey else null
         userService = RPCUserServiceImpl(configuration)
+
+        // Add System user permissions
+        defaultFlowWhiteList.forEach {
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: I think this will conflict with Shams' work - might want to rebase onto his.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 15]
```diff
@@ -121,6 +124,17 @@ class Node(override val configuration: FullNodeConfiguration, networkMapAddress:
...
+            userService.addSystemUserPermission(startFlowPermission(it.key))
+        }
+        pluginRegistries.forEach {
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: I know forEach can be convenient but I'm not a huge fan of the scope shadowing of `it` here. A good old fashioned for loop is basically a foreach in kotlin ...

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 16]
```diff
@@ -297,13 +306,20 @@ class Node(override val configuration: FullNodeConfiguration, networkMapAddress:
...
+        client.start(RPCUserService.systemUserUsername, RPCUserService.systemUserPassword)
+        return client.proxy()
+    }
+
     override fun start(): Node {
         alreadyRunningNodeCheck()
         super.start()
         // Only start the service API requests once the network map registration is complete
-        networkMapRegistrationFuture.then {
+        thread(name = "WebServer") {
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: Why put this into the background? I think it will mess up the measurement of startup time at least, but it seems an unrelated change?

File: node/src/main/kotlin/net/corda/node/services/RPCUserService.kt (hunk 5 above)
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/services/RPCUserService.kt]: Isn't this redundant after your change to use TLS certs for intra-node components, shams?

File: node/src/main/kotlin/net/corda/node/services/messaging/RPCStructures.kt [hunk 17]
```diff
@@ -218,6 +205,9 @@ private class RPCKryo(observableSerializer: Serializer<Observable<Any>>? = null)
         register(FlowHandle::class.java)
         register(KryoException::class.java)
         register(StringBuffer::class.java)
+        register(Unit::class.java)
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/services/messaging/RPCStructures.kt]: Are we really serialising Unit over the wire?!

File: samples/raft-notary-demo/src/main/kotlin/net/corda/notarydemo/api/NotaryDemoApi.kt [hunk 18]
```diff
@@ -3,28 +3,29 @@ package net.corda.notarydemo.api
...
 import javax.ws.rs.Path
 import javax.ws.rs.PathParam
 import javax.ws.rs.core.Response
 
 @Path("notarydemo")
-class NotaryDemoApi(val services: ServiceHub) {
+class NotaryDemoApi(val rpc: CordaRPCOps) {
     private val notary by lazy {
-        services.networkMapCache.getAnyNotary() ?: throw IllegalStateException("No notary found on the network")
+        rpc.networkMapUpdates().first.first { it.advertisedServices.any { it.info.type.isNotary() } }.notaryIdentity
```
Comment (mikehearn) [inline samples/raft-notary-demo/src/main/kotlin/net/corda/notarydemo/api/NotaryDemoApi.kt]: This sort of construct makes me think the rpc framework should support a notion of idempotent returns (i.e. an RPC always returns the same thing and can be de-duplicated on the client side).

File: samples/trader-demo/src/main/kotlin/net/corda/traderdemo/api/TraderDemoApi.kt [hunk 19]
```diff
@@ -29,34 +34,44 @@ class TraderDemoApi(val services: ServiceHub) {
     @Path("create-test-cash")
     @Consumes(MediaType.APPLICATION_JSON)
     fun createTestCash(params: TestCashParams): Response {
-        val notary = services.networkMapCache.notaryNodes.single { it.legalIdentity.name == params.notary }.notaryIdentity
-        services.fillWithSomeTestCash(params.amount.DOLLARS,
-                outputNotary = notary,
-                ownedBy = services.myInfo.legalIdentity.owningKey)
+        val notary = rpc.networkMapUpdates().first.first { it.legalIdentity.name == params.notary }
```
Comment (mikehearn) [inline samples/trader-demo/src/main/kotlin/net/corda/traderdemo/api/TraderDemoApi.kt]: This is such an ugly pattern .... it implies to me there should be a dedicated RPC for this common need.

File: node/src/main/kotlin/net/corda/node/services/RPCUserService.kt (hunk 5 above)
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/RPCUserService.kt]: Good spot! Have the webserver login to the Artemis server with the node user and because it's using the same keystore it will use the same TLS cert and get logged in as the node and get full permissions. The only thing missing would be to give full RPC permissions to the node user. I think that can easily fixed by updating `requirePermission`. I like it!

@exFalso my branch is shams-p2p-auth.

Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/RPCUserService.kt]: Ok, that sounds much cleaner, i'll rebase onto your branch

File: core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt [hunk 20]
```diff
@@ -112,6 +84,17 @@ interface CordaRPCOps : RPCOps {
      * Retrieve existing note(s) for a given Vault transaction
      */
     fun getVaultTransactionNotes(txnId: SecureHash): Iterable<String>
+
+    fun attachmentExists(id: SecureHash): Boolean
+    fun uploadAttachment(jar: InputStream): SecureHash
```
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt]: Yes, I was thinking we may need to expose some lower level api to stream efficiently.

I was also surprised, but it just worked out of the box

File: core/src/main/kotlin/net/corda/core/node/CordaPluginRegistry.kt [hunk 21]
```diff
@@ -1,52 +1,52 @@
...
  * to extend a Corda node with additional application services.
  */
-abstract class CordaPluginRegistry {
-    /**
-     * List of JAX-RS classes inside the contract jar. They are expected to have a single parameter constructor that takes a ServiceHub as input.
-     * These are listed as Class<*>, because in the future they will be instantiated inside a ClassLoader so that
-     * Cordapp code can be loaded dynamically.
-     */
-    open val webApis: List<Class<*>> = emptyList()
+abstract class CordaPluginRegistry(
```
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/node/CordaPluginRegistry.kt]: What do you mean? These used to be `val`s, I kept that, just moved them from the body of the class to it's constructor

Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/node/CordaPluginRegistry.kt]: Wrt lambdas, the api used to be `SimmFlow.Service::class.java`, this changed to `SimmFlow::Service`. The difference is that this does compile time checks, whereas the previous version did it with reflection, finding the first constructor that seemed good.

File: core/src/main/kotlin/net/corda/flows/FinalityFlow.kt [hunk 22]
```diff
@@ -16,7 +16,9 @@ import net.corda.core.utilities.ProgressTracker
  */
 class FinalityFlow(val transaction: SignedTransaction,
                    val participants: Set<Party>,
-                   override val progressTracker: ProgressTracker = tracker()) : FlowLogic<Unit>() {
+                   override val progressTracker: ProgressTracker) : FlowLogic<Unit>() {
+    constructor(transaction: SignedTransaction, participants: Set<Party>) : this(transaction, participants, tracker())
```
Comment (exFalso) [inline core/src/main/kotlin/net/corda/flows/FinalityFlow.kt]: Yes i tried that but it's not enough:S. Perhaps worth an issue raise @ jetbrains?

File: core/src/main/kotlin/net/corda/core/node/CordaPluginRegistry.kt [hunk 23]
```diff
@@ -1,52 +1,52 @@
...
+        /**
+         * List of lambdas returning JAX-RS objects. They may only depend on the RPC interface, as the webserver should
+         * potentially be able to live in a process separate from the node itself.
+         */
+        open val webApis: List<(CordaRPCOps) -> Any> = emptyList(),
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/CordaPluginRegistry.kt]: They are Java compatible but just that they are not the standard lambda interfaces that Java 8 uses.

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt [hunk 24]
```diff
@@ -84,8 +92,41 @@ class CordaRPCOpsImpl(
...
+                    id = stateMachine.id,
+                    flowLogicClassName = stateMachine.logic.javaClass.name,
+                    progressTrackerStepAndUpdates = stateMachine.logic.track()
+            )
+        }
+
+        fun stateMachineUpdateFromStateMachineChange(change: StateMachineManager.Change): StateMachineUpdate {
+            return when (change.addOrRemove) {
+                AddOrRemove.ADD -> {
+                    val stateMachineInfo = StateMachineInfo(
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: No, this is based on a state machine change, whereas the other is based on a statemachineimpl. I think the only code shared is `(..).logic.javaClass.name` and `(..).logic.track()`

File: node/src/main/kotlin/net/corda/node/services/messaging/RPCStructures.kt (hunk 17 above)
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/messaging/RPCStructures.kt]: Yes, there are flows that return Unit

File: samples/raft-notary-demo/src/main/kotlin/net/corda/notarydemo/api/NotaryDemoApi.kt (hunk 18 above)
Comment (exFalso) [inline samples/raft-notary-demo/src/main/kotlin/net/corda/notarydemo/api/NotaryDemoApi.kt]: You mean rpc memoization? Makes sense, but requires some thought e.g. it'd mean that from the first call on we'll hold a reference to the returned observable.

**Comment by mikehearn** [comment]
Is this ball in my court? I think it needs a rebase.

**Comment by exFalso** [comment]
This is now rebased, comments addressed. The NODE user handling is quite ugly, added TODOs to remove those pieces of code once the webserver doesnt need it

File: core/src/main/kotlin/net/corda/core/serialization/Kryo.kt [hunk 25]
```diff
@@ -202,6 +200,55 @@ class ImmutableClassSerializer<T : Any>(val klass: KClass<T>) : Serializer<T>()
...
 }
 
+// TODO This is a temporary inefficient serialiser for sending InputStreams through RPC. This may be done much more
+// efficiently using Artemis's large message feature.
+class InputStreamSerializer : Serializer<InputStream>() {
+    override fun write(kryo: Kryo, output: Output, stream: InputStream) {
+        val buffer = ByteArray(4096)
+        while (true) {
+            val numberOfBytesRead = stream.read(buffer)
+            if (numberOfBytesRead > 0) {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: -1 represents end of stream

File: core/src/main/kotlin/net/corda/core/serialization/Kryo.kt [hunk 26]
```diff
@@ -202,6 +200,55 @@ class ImmutableClassSerializer<T : Any>(val klass: KClass<T>) : Serializer<T>()
     }
 }
 
+// TODO This is a temporary inefficient serialiser for sending InputStreams through RPC. This may be done much more
+// efficiently using Artemis's large message feature.
+class InputStreamSerializer : Serializer<InputStream>() {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: object

File: core/src/main/kotlin/net/corda/core/serialization/Kryo.kt [hunk 27]
```diff
@@ -202,6 +200,55 @@ class ImmutableClassSerializer<T : Any>(val klass: KClass<T>) : Serializer<T>()
...
+        while (true) {
+            val chunk = input.readBytesWithLength()
+            if (chunk.isEmpty()) {
+                break
+            } else {
+                chunks.add(chunk)
+            }
+        }
+
+        return object : InputStream() {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: Why not flatten `chunks` and then use a `ByteArrayInputStream`?

File: node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt [hunk 28]
```diff
@@ -184,7 +184,9 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
         securityRoles["$INTERNAL_PREFIX#"] = setOf(nodeInternalRole)  // Do not add any other roles here as it's only for the node
         securityRoles[P2P_QUEUE] = setOf(nodeInternalRole, restrictedRole(PEER_ROLE, send = true))
         securityRoles[RPC_REQUESTS_QUEUE] = setOf(nodeInternalRole, restrictedRole(RPC_ROLE, send = true))
-        for ((username) in userService.users) {
+        // TODO remove NODE_USER once webserver doesn't need it
+        val possibleClientUserNames = userService.users.map { it.username } + listOf(NODE_USER)
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt]: Make it explicit for the NODE USER: `securityRoles["$CLIENTS_PREFIX$NODE_USER.rpc.*"] = setOf(nodeInternalRole)`

File: core/src/main/kotlin/net/corda/core/serialization/Kryo.kt (hunk 25 above)
Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: Yes

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: This will terminate on reading 0 bytes but that's not what the docs say. It only ends on -1.

Comment (exFalso) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: It can only return 0 if the input buffer is of size 0, which will not happen.
I'll just change it to `== -1`

File: node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt [hunk 29]
```diff
@@ -184,7 +184,10 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
         securityRoles["$INTERNAL_PREFIX#"] = setOf(nodeInternalRole)  // Do not add any other roles here as it's only for the node
         securityRoles[P2P_QUEUE] = setOf(nodeInternalRole, restrictedRole(PEER_ROLE, send = true))
         securityRoles[RPC_REQUESTS_QUEUE] = setOf(nodeInternalRole, restrictedRole(RPC_ROLE, send = true))
-        for ((username) in userService.users) {
+        // TODO remove the NODE_USER role once the webserver doesn't need it
+        securityRoles["$CLIENTS_PREFIX$NODE_USER.rpc.*"] = setOf(nodeInternalRole)
+        val possibleClientUserNames = userService.users.map { it.username }
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt]: Revert this change

File: node/src/main/kotlin/net/corda/node/services/messaging/RPCDispatcher.kt [hunk 30]
```diff
@@ -153,9 +160,19 @@ abstract class RPCDispatcher(val ops: RPCOps, val userService: RPCUserService) {
         return ClientRPCRequestMessage(SerializedBytes(argBytes), replyTo, observationsTo, methodName, user)
     }
 
+    // TODO remove this User once webserver doesn't need it
+    val nodeUser = User(NODE_USER, NODE_USER, setOf())
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/messaging/RPCDispatcher.kt]: You don't need to expose this as you're not using the object elsewhere.

**Comment by exFalso** [comment]
Comments addressed, @mikehearn is this good to go?

**Comment by mikehearn** [comment]
Yes, that looks OK now. The InputStream serialiser still looks weird - why bother chunking when you read the whole thing into RAM anyway - but as we'll toss it sooner or later it doesn't matter much.

The node internal user is not ideal: it'd be better to do least privilege for node internal components too, but that can be tackled as part of later security work. You could maybe add a brief discussion where NODE_USER is defined on the topic of privilege minimisation for node components. That'd be nice to have, to show readers of the code that we're thinking about it.
